x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -1587,4 +1589,72 @@ private static LogicalPlan normalize(Aggregate aggregate, AttributeMap<Expressio
             return changed.get() ? new Aggregate(aggregate.source(), aggregate.child(), aggregate.groupings(), newAggs) : aggregate;
         }
     }
+
+    public static class FoldNull extends OptimizerRules.OptimizerExpressionRule<Expression> {
+
+        public FoldNull() {
+            super(TransformDirection.UP);
+        }
+
+        @Override
+        protected Expression rule(Expression e) {
+            if (e instanceof In in) {
+                if (Expressions.isNull(in.value())) {
+                    return Literal.of(in, null);
+                }
+            } else if (e instanceof Alias == false
+                && e.nullable() == Nullability.TRUE
+                && Expressions.anyMatch(e.children(), Expressions::isNull)) {
+                    return Literal.of(e, null);
+                }
+            return e;
+        }
+    }
+
+    public static class PropagateNullable extends OptimizerRules.PropagateNullable {
+        @Override
+        protected Expression rule(And and) {
+            List<Expression> splits = Predicates.splitAnd(and);
+
+            Set<Expression> nullExpressions = new LinkedHashSet<>();
+            Set<Expression> notNullExpressions = new LinkedHashSet<>();
+            List<Expression> others = new LinkedList<>();
+
+            // first find isNull/isNotNull
+            for (Expression ex : splits) {
+                // Avoid considering constants. It would lead to collisions with foldable propagation
+                // Folding will take care of them eventually anyway
+                if (ex instanceof IsNull isn && isn.field().foldable() == false) {
+                    nullExpressions.add(isn.field());
+                } else if (ex instanceof IsNotNull isnn && isnn.field().foldable() == false) {
+                    notNullExpressions.add(isnn.field());
+                }
```

Comment:
Why is there a collision over foldables?

Reply:
I explained it here https://github.com/elastic/elasticsearch/pull/105691#discussion_r1512986433

The problem is that `PropagateEvalFoldables` (that only exists in ESQL, that is probably one of the reasons why we don't have this problem in SQL) replaces fields with their value when they are foldable, so:
- if `a = 1`, an expression like `a is null` could become `1 is null`
- for `a is null`, that now is `1 is null`, `isn.field()` returns `1`, that gets added to nullExpressions
- nullExpressions are then replaced with NULL in other expressions in the same AND block, so also completely unrelated expressions like `b > 1` could be impacted, that is exactly the problem @astefan reported

In SQL the problem doesn't happen also because `nullify()` can rely on `nullable() = false` to filter out these cases. Paradoxically, if we cannot rely on the nullability logic in ESQL, we don't even know that `1` is nullable.

Reply:
> Paradoxically, if we cannot rely on the nullability logic in ESQL, we don't even know that 1 is not nullable.

in this specific case I could fold, but we have other rules that do that, this one is about nullability